datalad_next/utils/shell_connection.py

```diff
@@ -0,0 +1,305 @@
+from typing import (
+    Generator,
+    Iterable,
+)
+from zlib import decompressobj
+
+from more_itertools import consume
+
+from datalad_next.itertools import align_pattern
+from datalad_next.runners.iter_subproc import (
+    OutputFrom,
+    iter_subproc,
+)
+
+
+lgr = logging.getLogger('datalad.ext.next.utils.ssh_shell')
+
+
+@contextmanager
+def shell_connection(shell_cmd: list[str],
+                     ) -> Generator[ShellCommandExecutor, None, None]:
+    """Context manager that provides an interactive connection to a shell
+
+    This context manager uses the provided argument ``shell_cmd`` to start a
+    shell-subprocess. Usually the commands provided in ``shell_cmd`` will
+    start a client for a remote shell, e.g. ``ssh``.
+
+    :func:`shell_connection` returns an instance
+    of :class:`ShellCommandExecutor` in the ``as``-variable. This instance can
+    be used to interact with the shell. That means, it can be used to execute
+    commands to the shell, receive the data that the commands write to their
+    ``stdout``, and retrieve the return code of the executed commands. All
+    commands that are executed via the returned instance of
+    :class:`ShellCommandExecutor` are executed in the same shell.
+
+    Parameters
+    ----------
+    shell_cmd : list[str]
+        The command to execute the shell. It should be a list of strings that
+        is given to :func:`iter_subproc` as `args`-parameter. For example:
+        ``['ssh', '-p', '2222', 'localhost']``.
+
+    Yields
+    ------
+    :class:`ShellCommandExecutor`
+    """
+
+    def train(queue: Queue):
+        """Use a queue to allow chaining of iterables at different times"""
+        for iterable in iter(queue.get, None):
+            yield from iterable
+
+    subprocess_inputs: Queue = Queue()
+    with iter_subproc(shell_cmd,
+                      input=train(subprocess_inputs),
+                      bufsize=0) as shell_output:
+        cmd_executor = ShellCommandExecutor(subprocess_inputs, shell_output)
+        try:
+            # Skip initial login messages
+            result_0 = cmd_executor.execute_command(b'test')
+            for line in result_0:
+                lgr.debug('skipped login message line: %s', line)
+            if result_0.returncode != 1:
+                raise RuntimeError(
+                    'unexpected return code from command'
+                    f'`test`: {result_0.returncode}')
+
+            # Return the now ready connection
+            yield cmd_executor
+        finally:
+            # Ensure that ssh it terminated if an exception is raised by code
+            # that uses `ssh_connection`. This is necessary because
+            # `terminate`-call that is invoked by leaving the
+            # `iterable_subprocess`-context will not end the ssh-process. It
+            # will only terminate if stdin is closed, or if it is killed.
+            subprocess_inputs.put(None)
+
+
+class ShellCommandExecutor:
```

Comment:
Should get a `__repr__`. Presently, it is not possible to say what shell one is dealing with when debugging:

```
(Pdb) p shell
<datalad_next.utils.shell_connection.ShellCommandExecutor object at 0x7f22825f2290>
```

Might be worth adding the actual "connection" args to the class for that, or support a dedicated label generated from them.

Reply:
Done in 3f4d884d1531140d90f8890697336160a2696aba